### ISSUE TITLE
`:Telescope lsp_workspace_diagnostics` causes `Invalid buffer id`

### ISSUE THREAD

**Issue by statiolake** [nvim-telescope/telescope.nvim#910]
### Description

`:Telescope lsp_workspace_diagnostics` sometimes causes error `Invalid buffer id`:

```
E5108: Error executing lua ...nfig/nvim/plugged/telescope.nvim/lua/telescope/utils.lua:198: Invalid buffer id: 1
```

`lua/telescope/utils.lua:198` is here,

https://github.com/nvim-telescope/telescope.nvim/blob/398a0d391aa2afcda1521a8e86b820ffb599b60b/lua/telescope/utils.lua#L198

and this is called from here.

https://github.com/nvim-telescope/telescope.nvim/blob/398a0d391aa2afcda1521a8e86b820ffb599b60b/lua/telescope/utils.lua#L218-L229

In the above, it seems that sometimes `vim.lsp.diagnostic.get_all()` returns invalid buffer ids. I don't know when exactly the invalid buffer ids are returned, but `:bwipeout`ing another buffer frequently causes this (see Reproduce section). From what little I've tried, It seems that this error can be avoided by checking if `bufnr` is valid before inner loop, like this:

```lua
for bufnr, diags in pairs(buffer_diags) do
  if vim.api.nvim_buf_is_valid(bufnr) then -- <- added this check
    for _, diag in ipairs(diags) do
      -- workspace diagnostics may include empty tables for unused bufnr
      if not vim.tbl_isempty(diag) then
        if filter_diag_severity(opts, diag.severity) then
          table.insert(items, preprocess_diag(diag, bufnr))
        end
      end
    end
  end
end
```

**Expected Behavior**

The picker opens and lists all workspace diagnostics.

**Actual Behavior**

`Invalid buffer id` error sometimes occurs and picker won't open until I restart entire Neovim.

### Details

I took a screen record to show what happens if you follow the steps (after opening Neovim) in the Reproduce section.

<details><summary>Record</summary>

https://user-images.githubusercontent.com/20490597/121816215-5446b900-ccb5-11eb-94da-864e963bd73a.mp4

</details>

<details><summary>Reproduce</summary>

1. Prepare Rust development environment and Rust Analyzer.
1. Create some working directory and `cd` to it.
1. Generate new cargo project by `cargo new testcrate`
1. `cd testcrate`
1. Replace auto-generated `src/main.rs` to the following contents. The string "some syntax error here" is not a valid Rust syntax, so Rust Analyzer will produce an error in later step.
  ```rust
  mod foo;

  fn main() {
      some syntax error here
  }
  ```
1. Create new file at `src/foo.rs` with the following contents. The string "another syntax error" is not a valid Rust syntax, so error in later step too.
  ```rust
  fn bar() {
      another syntax error
  }
  ```
1. (Above recording starts from here) Open neovim by `nvim -nu test.vim`.
1. (You may need to restart `nvim -nu test.vim` to use plugged nvim-lspconfig)
1. `:e src/main.rs` and you'll see errors shown on the line by Neovim's virtual text.
1. `:e src/foo.rs` and you'll see errors too.
1. `:Telescope lsp_workspace_diagnostics` and you'll see workspace errors. While you're opening `src/foo.rs` but you'll also see errors on `src/main.rs`. If the errors in `src/main.rs` are not shown, restarting `:Telescope lsp_workspace_diagnostics`, saving the buffer, or changing files may help.
1. After you confirmed the errors on both files are listed, close Telescope by `<Esc>`.
1. Wipeout the hidden (not showing) buffer, `src/main.rs`, by `:bwipeout src/main.rs`. In the above screen record, I'm executing `:ls` and then `:bwipeout {bufnr}`, but either is OK.
1. Try `:Telescope lsp_workspace_diagnostics` again. Probably you'll see the `Invalid buffer id` error and the picker doesn't show up. Note that sometimes it will be correctly shown.
1. You can't open `:Telescope lsp_workspace_diagnostics` anymore. In the above screen record, I reopened wipeouted `src/main.rs` again, but that doesn't make it working again.
1. If you are in that situation, probably you'll find the wipeouted buffer number in `vim.lsp.diagnostic.get_all()`.

</details>

<details><summary>Environment</summary>

- nvim --version output: 
    ```
    NVIM v0.5.0-dev+1321-gc57a85d53
    Build type: RelWithDebInfo
    LuaJIT 2.0.5
    Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions     -Wp,-D_FORTIFY_SOURCE=2,-D_GLIBCXX_ASSERTIONS     -Wformat -Werror=format-security     -fstack-clash-protection -fcf-protection -Wp,-U_FORTIFY_SOURCE -Wp,-D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/tmp/pakku-dicen/neovim-nightly-git/src/build/config -I/tmp/pakku-dicen/neovim-nightly-git/src/neovim-nightly-git/src -I/usr/include -I/tmp/pakku-dicen/neovim-nightly-git/src/build/src/nvim/auto -I/tmp/pakku-dicen/neovim-nightly-git/src/build/include
    Compiled by dicen

    Features: +acl +iconv +tui
    See ":help feature-compile"

          システム vimrc: "$VIM/sysinit.vim"
           省略時の $VIM: "/usr/share/nvim"

    Run :checkhealth for more info
    ```
- Operating system: Recorded in Arch Linux but I also confirmed on Windows
- Telescope commit: '398a0d3'

</details>

<details><summary>Configuration</summary>
<p>

```viml
set nocompatible hidden laststatus=2

if !filereadable('/tmp/plug.vim')
  silent !curl --insecure -fLo /tmp/plug.vim
    \ https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim
endif

source /tmp/plug.vim
call plug#begin('/tmp/plugged')
Plug 'nvim-lua/popup.nvim'
Plug 'nvim-lua/plenary.nvim'
Plug 'nvim-telescope/telescope.nvim'
Plug 'neovim/nvim-lspconfig'
call plug#end()

autocmd VimEnter * PlugClean! | PlugUpdate --sync | close

lua << EOF

-- change path to rust-analyzer if needed
require('lspconfig').rust_analyzer.setup {
  cmd = {'rust-analyzer-linux'},
}

require('telescope').setup{
  defaults = {
    vimgrep_arguments = {
      'rg',
      '--color=never',
      '--no-heading',
      '--with-filename',
      '--line-number',
      '--column',
      '--smart-case'
    },
    prompt_position = "bottom",
    prompt_prefix = ">",
    selection_strategy = "reset",
    sorting_strategy = "descending",
    layout_strategy = "horizontal",
    layout_defaults = {},
    file_ignore_patterns = {},
    shorten_path = true,
    winblend = 0,
    width = 0.75,
    preview_cutoff = 120,
    results_height = 1,
    results_width = 0.8,
    border = {},
    borderchars = { '─', '│', '─', '│', '╭', '╮', '╯', '╰'},
    color_devicons = true,
    use_less = true,
  }
}

```

**Comment by fdschmidt93** [comment]
Many thanks for the detailed report! It's easy to the reproduce the issue :)

Also many thanks for coming up with an initial solution! I've tried it and it generally works; however, a slight issue appears I also haven't yet figured out: once you `:bwipeout` a buffer, it seems that it won't reappear on `lsp_workspace_diagnostics`. Ideally, our solution can address both the error and allow for the wiped out buffer to appear in `lsp_workspace_diagnostics` regardless.

I'll try to look into it some time this week, but I'd like to encourage you to also find a solution and submit a PR if you can :)

**Comment by statiolake** [comment]
Thank you for your quick reply :)

> once you `:bwipeout` a buffer, it seems that it won't reappear on` lsp_workspace_diagnostics`

Actually it reappears if you explicitly reopen the wiped out file. However, if you look at the output of `vim.lsp.diagnostic.get_all()` after reopening, the old buffer number still persists in the table but new entry is also created for a reopened buffer. New diagnostics caused by editing the reopened file is only added to the table associated with the reopened buffer. Old entry will never be updated. So I think the root cause of this bug is not a telescope, but rather the way Neovim handles LSP diagnostics. I wonder this Neovim's behavior is expected or not.

Perhaps we can workaround this by tracking each buffer's file names separately (such as storing values on other table), but handling old diagnostics left in the old entry will be very hard. The old entry will never be updated so the entry keep all of old diagnostics even if actually the diagnostics are all gone by editing in the reopened buffer. The entry for the reopened file is correctly updated, so it can also workarounded by tracking latest buffer numbers for each file. However, finding the buffer previously opening the file seems almost impossible: simply by comparing file name won't work when file is renamed before reopening. Considering all of those, I think this requires too much effort to workaround. At least I don't have any good ideas for now...

For example:

1. Suppose `a.rs` (buffer 1) and `b.rs` (buffer 2) both have errors, `err_a` and `err_b` respectively:
    → `vim.lsp.diagnostic.get_all()` would be something like `{ [1] = { err_a }, [2] = { err_b } }`. The telescope picker shows `err_a` and `err_b`.
1. Wipe out the buffer `b.rs`:
    → `vim.lsp.diagnostic.get_all()` would be the same with the above.
    - Current telescope can't open the picker in this situation.
    - If we check only for the buffer validity, the picker would show `err_a`. (not ideal)
    - If we tracked buffer name to work around, the picker would show `err_a` and `err_b`. (ideal)
1. Rename `b.rs` to `c.rs`, and open it (buffer 3):
    → `vim.lsp.diagnostic.get_all()` would be something like `{ [1] = { err_a }, [2] = { err_b }, [3] = { err_b } }`.
    - Current telescope still doesn't open.
    - If we check only for the buffer validity, the picker would show `err_a` and `err_b` for `c.rs`. (ideal)
    - If we tracked buffer name to work around, `err_a`, `err_b` for `c.rs`, and `err_b` for already renamed `b.rs` would be shown. (not ideal)
      We should ignore this old `b.rs` diagnostic, but we need to distinguish whether a newly opened `c.rs` is a renamed version of a previously opened file or a truly new file in order to do that. Difficult.
1. Solve `err_b` by editing `c.rs` (buffer 3)
    → `vim.lsp.diagnostic.get_all()` would be something like `{ [1] = { err_a }, [2] = { err_b }, [3] = {} }`.
    - Current telescope still doesn't open.
    - If we check only for the buffer validity, the picker would show `err_a`. (ideal)
    - If we tracked buffer name to work around, `err_a` and `err_b` for `b.rs` would be shown. (not ideal)

In conclusion, I don't know when the cached file names should be invalidated, i.e. when the old diagnostics should be ignored. I also think those are an LSP client's job. At least it's too difficult for me... Any thoughts?

**Comment by fdschmidt93** [comment]
> Actually it reappears if you explicitly reopen the wiped out file. However, if you look at the output of vim.lsp.diagnostic.get_all() after reopening, the old buffer number still persists in the table but new entry is also created for a reopened buffer. New diagnostics caused by editing the reopened file is only added to the table associated with the reopened buffer. Old entry will never be updated. So I think the root cause of this bug is not a telescope, but rather the way Neovim handles LSP diagnostics. I wonder this Neovim's behavior is expected or not.

I referred to the case right after wiping out :) but I saw it appears as a "new buffer" once the buffer is reopened (assuming your fix is applied).

I'd agree, my hunch is this is more an upstream issue, as in, for instance, should workspace diagnostics rather be tracked by URI as opposed to buffer number. In such a case, funny stuff might happen _regardless_. Even more so, I suppose some junks of telescope jump-to entry functionality (ie hit enter on a diagnostic) would not work out-of-the-box and be adapted as well to the case the referring buffer got wiped out.

My hunch is the cleanest solution would, if feasible, would be to reload diagnostics _entirely_ (a hook triggered by change in "buffer  namespace" or by a file watcher), as in; you wipe out a buffer or one of the files in the LSP diagnostics was changed, and upon the next call of `vim.lsp.diagnostic.get_all` all background buffers exclusively relating to diagnostics would be cleared and their buffers reinstantiated. Since this is probably something that should be implemented upstream `(`vim.lsp.diagnostic.get_all()`), I'd raise an issue upstream (or even better a PR :)  in your shoes. While this seems like a good feature to have in an LSP client it's quite the edge case, for instance, I'm afraid I'm not sure in what scenarios I'd have to `bwipeout` a buffer to begin with.

**Comment by statiolake** [comment]
I noticed the fix was already in Neovim: https://github.com/neovim/neovim/pull/14676.
